lightning/src/ln/channelmanager.rs

```diff
@@ -422,6 +422,11 @@ pub struct ChannelDetails {
 	pub short_channel_id: Option<u64>,
 	/// The node_id of our counterparty
 	pub remote_network_id: PublicKey,
+	/// The Features this node provided us opon our last connection with them.
+	/// This is particularly useful for routing as many node_announcement-context features are also
+	/// set in the init-context, and we should almost certainly consider the init-context version
+	/// to be the latest copy.
+	pub remote_node_features: InitFeatures,
```

Comment:
I'm not sure about assuming the init-context version is the latest copy, a peer may be able to change its advertised services without renewing any connection with us, just by broadcasting a new node_announcement, so between init and node_announcement I think we should reconciliate features where we have both states available, i.e in get_route ? 

But do we really need to cache them in ChannelManager, can't we store everything in PeerManager ? It's harder to see given a lot of features are obviously yet to be designed, so maybe we should have at least an access function returning NodeDetails to let client select nodes.

Reply:
Hmm, I was kinda assuming to get new features you'd at least have to restart your node. It may make sense to further clarify the BOLTs in this regard, but certainly today the restart requirement holds (and it may be some time after restart before you create a new announcement, so Init would be the latest-correct).

As for storing it in peer_manager, we could do that, but it just makes for an awkward API - it would mean the client has to call something on ChannelManger, get a list of channels, then go ask PeerHandler for the InitFeatures for each node that we are connected to. It's not exactly an obviously-correct place to store it, but I don't think its completely absurd for ChannelManager to store the latest InitFeatures we have, especially given it already tracks connected-ness state for peers.

Reply:
I thought at first that features were considered the same than service bits on base layer and so stable until restart but I can't find this explicit requirement in the bolts, maybe we should amend them ?

Based on this we may have to adapt the API to conciliate both messages but if your interpretation holds I'm okay with the API.

The NodeDetails is a different point though and I think it's needed anyway (e.g a client don't want to open channel with nodes not using opt_anchor_output or other), can be added in a follow-up PR.

Reply:
Right, its definitely not "required" by the BOLTs but I figured it was a safe assumption. As for init features overriding node_announcement features, I think thats also a reasonable assumption but I agree that we should make that explicit in the BOLTs.